aggregator/entry.go

```diff
@@ -173,52 +234,102 @@ func (e *Entry) TryExpire(now time.Time) bool {
 }
 
 func (e *Entry) writerCount() int        { return int(atomic.LoadInt32(&e.numWriters)) }
-func (e *Entry) lastAccessed() time.Time { return time.Unix(0, atomic.LoadInt64(&e.lastAccessInNs)) }
+func (e *Entry) lastAccessed() time.Time { return time.Unix(0, atomic.LoadInt64(&e.lastAccessNanos)) }
 
 func (e *Entry) recordLastAccessed(currTime time.Time) {
-	atomic.StoreInt64(&e.lastAccessInNs, currTime.UnixNano())
+	atomic.StoreInt64(&e.lastAccessNanos, currTime.UnixNano())
 }
 
-func (e *Entry) shouldExpire(now time.Time) bool {
-	// Only expire the entry if there are no active writers
-	// and it has reached its ttl since last accessed.
-	return e.writerCount() == 0 && now.After(e.lastAccessed().Add(e.opts.EntryTTL()))
+// NB(xichen): we assume the policies are sorted by their cutover times
+// in ascending order.
+func (e *Entry) activeStagedPoliciesWithLock(
+	pl policy.PoliciesList,
+	t time.Time,
+) (policy.StagedPolicies, error) {
+	// If we have no policy to apply, simply bail.
+	if len(pl) == 0 {
+		e.metrics.emptyPoliciesList.Inc(1)
+		return policy.DefaultStagedPolicies, errEmptyPoliciesList
+	}
+	timeNanos := t.UnixNano()
+	for idx := len(pl) - 1; idx >= 0; idx-- {
+		if pl[idx].CutoverNanos <= timeNanos {
+			return pl[idx], nil
+		}
+	}
+	return pl[0], nil
```

Comment:
This is the case where all policies are in the future, in which case we'll use the cached policy instead. If we use a default policy here, it may give us the wrong results.